### PR TITLE
Dictionary deinflections

### PR DESCRIPTION
**Pull request by StefanVukovic99**
Supports dictionaries containing deinflection information that combines with the deinflector's. To accommodate this, some changes are made to how deinflection information is structured and shown. 

Incidentally, some Japanese deinflections that seem to have been missing are now shown, like the second one in the 読め example:
image

Part 5 of #422.

### CODE REVIEW

File: ext/js/language/translator.js [hunk 1]
```diff
@@ -236,49 +248,153 @@ export class Translator {
...
+    /**
+     * @param {import('dictionary').Inflections} hypothesis1
+     * @param {import('dictionary').Inflections} hypothesis2
+     * @returns {boolean}
+     */
+    _areInflectionHyphothesesEqual(hypothesis1, hypothesis2) {
+        const set1 = new Set(hypothesis1);
+        const set2 = new Set(hypothesis2);
+
+        return set1.size === set2.size && [...set1].every((x) => set2.has(x));
```
Comment (toasted-nutbread) [inline ext/js/language/translator.js:278]: This is probably more efficient if you only construct 1 set and no temporary array. Also do the size comparison before any of that for an earlier exit.

File: ext/js/language/translator.js [hunk 2]
```diff
@@ -236,49 +248,153 @@ export class Translator {
...
+    /**
+     * @param {import('translation-internal').DatabaseDeinflection[]} deinflections
+     * @param {Map<string, import('translation').FindTermDictionary>} enabledDictionaryMap
+     * @param {import('dictionary').TermSourceMatchType} matchType
+     * @returns {Promise<import('translation-internal').DatabaseDeinflection[]>}
+     */
+    async _getDictionaryDeinflections(deinflections, enabledDictionaryMap, matchType) {
+        /** @type {import('translation-internal').DatabaseDeinflection[]} */
+        const dictionaryDeinflections = [];
+        deinflections.forEach((deinflection) => {
```
Comment (toasted-nutbread) [inline ext/js/language/translator.js:322]: don't use forEach, use `for of` loop

File: ext/js/language/translator.js [hunk 3]
```diff
@@ -236,49 +248,153 @@ export class Translator {
...
+
+    /**
+     * @param {import('translation-internal').DatabaseDeinflection[]} deinflections
+     * @param {Map<string, import('translation').FindTermDictionary>} enabledDictionaryMap
+     * @param {import('dictionary').TermSourceMatchType} matchType
+     * @param {boolean} partsOfSpeechFilter
+     */
+    async _addEntriesToDeinflections(deinflections, enabledDictionaryMap, matchType, partsOfSpeechFilter = true) {
+        const uniqueDeinflectionsMap = this._groupDeinflectionsByTerm(deinflections);
+        const uniqueDeinflectionArrays = Array.from(uniqueDeinflectionsMap.values());
```
Comment (toasted-nutbread) [inline ext/js/language/translator.js:361]: Use spread instead of Array.from
`[...something.values()]`

File: ext/js/language/translator.js [hunk 4]
```diff
@@ -236,49 +248,153 @@ export class Translator {
...
         const databaseEntries = await this._database.findTermsBulk(uniqueDeinflectionTerms, enabledDictionaryMap, matchType);
+        this._matchEntriesToDeinflections(databaseEntries, uniqueDeinflectionArrays, partsOfSpeechFilter);
+    }
+
+    /**
+     * @param {import('translation-internal').DatabaseDeinflection[]} deinflections
+     * @returns {Map<string, import('translation-internal').DatabaseDeinflection[]>}
+     */
+    _groupDeinflectionsByTerm(deinflections) {
+        const result = deinflections.reduce((map, deinflection) => {
```
Comment (toasted-nutbread) [inline ext/js/language/translator.js:373]: IMO this is more clear and consistent with other parts of the codebase if you don't use reduce.

```js
/** @type {Map<string, import('translation-internal').DatabaseDeinflection[]>} */
const result = new Map();
for (const deinflection of deinflections) {
    const {deinflectedText} = deinflection;
    let deinflectionArray = result.get(deinflectedText);
    if (typeof deinflectionArray === 'undefined') {
        deinflectionArray = [];
        result.set(deinflectedText, deinflectionArray);
    }
    deinflectionArray.push(deinflection);
}
return result;
```
(this impl also avoids redundant `.set` which is also more consistent with other parts of translator)

File: ext/js/language/translator.js (hunk 1 above)
Comment (StefanVukovic99) [reply ext/js/language/translator.js]: On closer inspection, what I actually need here is something like 
```js
DictionaryDataUtil.areArraysEqual(
    [...hypothesis.inflections].sort(),
    [...inflections].sort()
));
```
is that ok?

Comment (toasted-nutbread) [reply ext/js/language/translator.js]: I'd say that's probably still more inefficient than it needs to be, esp. given that these arrays are probably likely to just be 0-2 long in the normal case. I'd suggest creating some utility function in core or a new core/*.js file that's like `areArraysEqualIgnoreOrder` and just do a 2-deep for loop.

```js
export function areArraysEqualIgnoreOrder(array1, array2) {
    const {length} = array1;
    if (length !== array2.length) { return false; }
    for (let i = 0; i < length; ++i) {
        if (!array2.includes(array1[i])) { return false; }
    }
    return true;
}
```

Rationale: for the expected use case of very small arrays, just doing what an in-place check is likely faster than reconstructing a new object, be it a set or a sorted array. I've always had some emphasis on keeping the translator fairly efficient, which is driving this feedback.

File: ext/js/dictionary/dictionary-data-util.js [hunk 5]
```diff
@@ -351,7 +351,7 @@ export class DictionaryDataUtil {
      * @param {T[]} array2
      * @returns {boolean}
      */
-    static _areArraysEqual(array1, array2) {
```
Comment (Scrub1492) [inline ext/js/dictionary/dictionary-data-util.js:354]: Why did you remove the underscore prefix?

File: ext/js/display/display-generator.js [hunk 6]
```diff
@@ -356,6 +356,46 @@ export class DisplayGenerator {
...
+
+    /**
+     * @param {import('dictionary').InflectionSource} source
+     * @returns {HTMLElement}
+     */
+    _getInflectionSourceIcon(source) {
+        const icon = document.createElement('span');
+        icon.style.marginRight = '0.5em';
+        switch (source) {
+            case 'dictionary':
```
Comment (Scrub1492) [inline ext/js/display/display-generator.js:384]: Just a suggestion, but is there any way the user can change the icons?

File: ext/js/dictionary/dictionary-data-util.js (hunk 5 above)
Comment (StefanVukovic99) [reply ext/js/dictionary/dictionary-data-util.js:354]: I used it outside the class so it's no longer "private".

File: ext/js/language/translator.js (hunk 1 above)
Comment (StefanVukovic99) [reply ext/js/language/translator.js]: This is an edge case, not sure if it might come up for some language, but  `areArraysEqualIgnoreOrder(["a", "b", "b"], ["a", "a", "b"])` is `true`.

File: ext/js/display/display-generator.js (hunk 6 above)
Comment (StefanVukovic99) [reply ext/js/display/display-generator.js:384]: Should be doable with custom css. Would also be nice to have actual icons.

File: ext/js/dictionary/dictionary-data-util.js (hunk 5 above)
Comment (Scrub1492) [reply ext/js/dictionary/dictionary-data-util.js:354]: Is it in this pr? I see no references outside of the class:

File: ext/js/language/translator.js (hunk 1 above)
Comment (StefanVukovic99) [reply ext/js/language/translator.js]: Actually, this seems to be the most performant implementation:
```js
   function areArraysEqualIgnoreOrder(array1, array2) {
        if (array1.length !== array2.length) {
            return false;
        }
    
        const frequencyCounter = {};
    
        for (const element of array1) {
            frequencyCounter[element] = (frequencyCounter[element] || 0) + 1;
        }
    
        for (const element of array2) {
            if (!frequencyCounter[element]) {
                return false;
            }
            frequencyCounter[element]--;
        }
       
        return true;
    }  
```

File: ext/js/dictionary/dictionary-data-util.js (hunk 5 above)
Comment (StefanVukovic99) [reply ext/js/dictionary/dictionary-data-util.js:354]: It's in `translator.js`, though we'll probably change that back and use a different function (see above)

File: ext/js/language/translator.js (hunk 1 above)
Comment (toasted-nutbread) [reply ext/js/language/translator.js]: How does that compare to using a map instead of `{}`?

Comment (toasted-nutbread) [reply ext/js/language/translator.js]: That implementation also only works for strings and has some edge cases related to `Object.prototype`, e.g. an element named `toString` will break it.

Comment (StefanVukovic99) [reply ext/js/language/translator.js]: I had made a basic [benchmark](https://jsbench.me/r8lr29janj/1) and it kept telling me Map is slower, but that might be an artifact of the setup. Map does seem better for this.

Comment (toasted-nutbread) [reply ext/js/language/translator.js]: While I do advocate for performance considerations where appropriate, I wouldn't want to sacrifice good code in favor of it; IMO object-as-a-map is not good modern JS. (I mentioned my thought about that in [point 1 here](https://github.com/themoeway/yomitan/issues/342#issue-2029729570).)

Comment (toasted-nutbread) [reply ext/js/language/translator.js]: Another question that comes to mind: are arrays with repeats valid for your use case? Like should those ever be generated? I know there are several situations where the extension generates an "array", but implicitly each item is unique so it's only an "array" for serialization purposes.

If each element is expected to be unique, which I think was my original assumption, you can probably simplify the implementation (e.g. change Map to Set). If not, I think the length equality check needs removal if you want to consider `["a", "b", "b"] = ["a", "a", "b"]` and `["a", "b", "b"] = ["a", "b"]`.

File: ext/js/display/display-generator.js (hunk 6 above)
Comment (toasted-nutbread) [reply ext/js/display/display-generator.js:384]: Going to revive this; can you make it so that icon has no `textContent` and instead has a `dataset.` property which specifies the type, and the content is driven by a CSS `::after` property? This will make it easier to style and also make it so the icon isn't selectable.

File: types/ext/dictionary.d.ts [hunk 7]
```diff
@@ -210,7 +210,7 @@ export type TermDictionaryEntry = {
     /**
      * A list of inflections that was applied to get the term.
```
Comment (djahandarie) [inline types/ext/dictionary.d.ts:211]: Is this still the best description? It seems to not reflect the "hypothesis" concept. Also I wonder if "hypothesis" is the best word, as that is something I would expect more in some sort of logical deduction system. Here I wonder if it's just more `possibleInflections`, `potentialInflections`, `inflectionCandidates` or something like that?

File: types/ext/dictionary.d.ts [hunk 8]
```diff
@@ -253,6 +253,15 @@ export type TermDictionaryEntry = {
     frequencies: TermFrequency[];
 };
 
+export type InflectionHypothesis = {
```
Comment (djahandarie) [inline types/ext/dictionary.d.ts:256]: Is it really a "hypothesis" at this stage, or is this just a grouped/labeled set of inflections?

**Review by djahandarie — COMMENTED** [review]
LGTM aside from variable description and naming (though it's totally possible I just don't understand what's going on).

File: types/ext/dictionary.d.ts (hunk 7 above)
Comment (StefanVukovic99) [reply types/ext/dictionary.d.ts]: I was hoping someone would suggest something about that. What I'm going for is: 
- give the individual, atomic inflections a different name from the composite 
- not imply that the list of composite inflections is exhaustive

I've changed it to "inflectionPossibility" (the strings that make up an inflectionPossibility can be "inflections"), and updated the description. Better?

**Comment by StefanVukovic99** [comment]
I've added a test case.

File: types/ext/dictionary.d.ts (hunk 7 above)
Comment (djahandarie) [reply types/ext/dictionary.d.ts]: Hmm, I think we need to think more carefully about this.

I think we have the following concepts:
* Inflection Rule: A single atomic operation that can be performed on a word to inflect it.
* Inflection Rule Chain: Multiple inflection rules applied in order. (I think it's not really a "possibility" at this point, that's more a function of how the layer above it uses this type.)
* Potential Inflection Rule Chains: A list of Inflection Rule Chains that are ways that a looked-up word can be reached from a dictionary headword.

And thinking from first principles about how this should be handled in code:
* "Inflection Rule" should almost certainly be a type (`InflectionRule`, or just `Rule` if we are in an appropriately named module), since it's very clearly a union of a couple different pieces of information. I would imagine the "source" should be on this type, since one could imagine building an Inflection Rule Chain from multiple Inflection Rules sourced from different places (although I guess your code doesn't currently do this).
* An "Inflection Rule Chain" could honestly just be a list, I'm not sure it really needs its own type (`InflectionRuleChain`), but I guess it would be okay as such to add a bit of semantic clarity and avoid `[][]` in the following bullet.
* "Potential Inflection Rule Chains" honestly feels like it should just be a variable (`potentialInflectionRuleChains`) pointing to the type `InflectionRuleChain[]` (or `Set<InflectionRuleChain>` since the order doesn't matter) at the location of whatever is using Inflection Rule Chains in a certain way.

Some things I considered while writing the above:
* I think "Inflection" alone is way too ambiguous whether it's referring to a process, a rule, a chain of roles, or the resulting word of after applying a rule. IMO it's best to have it refer to the overall process/concept, and be more explicit about calling things rules or results.
* I had to consider whether "deinflection" was more appropriate, but again, I think it's best that it refers to the process  (as opposed to rules, chains of rules, or a result of applying that chain). Also, since the same concepts would be used in inflection and deinflecting processes, I think the types can simply just refer to "inflection" since that is the more core concept (as opposed to the negated concept of deinflection). Then the module code and/or function names can of course refer to "deinflection" if that is what they are doing.
* We should be careful to support both agglutinative and fusional languages here. AFAIK, both of those use the concept of inflection, but fusional languages will just have shorter chains (usually of length 1 probably). I think that fits fine in the above breakdown. I suppose we also need to support analytic languages but I think that is not a problem as they will simply always be direct dictionary hits and avoid the deinflector entirely.
* We need to be careful to avoid the term "derivation" since that is a similar but separate concept in linguistics.

(This is just meant to be ground-up thinking as opposed to a comment on your PR. Feel free to take this and integrate it however you think it fits best. Also cc @toasted-nutbread)

File: ext/js/display/display-generator.js [hunk 9]
```diff
@@ -356,6 +356,46 @@ export class DisplayGenerator {
...
+        return fragment;
+    }
+
+    /**
+     * @param {import('dictionary').InflectionSource} source
+     * @returns {HTMLElement}
+     */
+    _getInflectionSourceIcon(source) {
+        const icon = document.createElement('span');
+        icon.classList.add('inflection-source-icon');
```
Comment (toasted-nutbread) [inline ext/js/display/display-generator.js:382]: Simplify 3 assignments to just `icon.dataset.inflectionSource = source;` outside of the switch.

File: types/ext/dictionary.d.ts (hunk 7 above)
Comment (StefanVukovic99) [reply types/ext/dictionary.d.ts]: Lots of good points, thanks. I've kept `PossibleInflectionRuleChain` as a type for now because the source property is on it, and `InflectionRuleChain` as a string[]. Other than that, the nomenclature is now pretty much as you described it.

File: ext/data/templates/default-anki-field-templates.handlebars
Comment (toasted-nutbread) [inline ext/data/templates/default-anki-field-templates.handlebars:1]: Since this file was updated, it probably needs to be updated in the settings upgrade as well.

File: ext/js/language/translator.js [hunk 10]
```diff
@@ -19,7 +19,6 @@
 import {RegexUtil} from '../general/regex-util.js';
 import {TextSourceMap} from '../general/text-source-map.js';
 import {Deinflector} from './deinflector.js';
-
```
Comment (toasted-nutbread) [inline ext/js/language/translator.js:22]: Preserve whitespace after imports.

File: types/ext/dictionary.d.ts (hunk 7 above)
Comment (toasted-nutbread) [reply types/ext/dictionary.d.ts]: I'm a bit late getting back to this, but overall I agree with the sentiment. I've had some thoughts about the deinflector class and its types being kinda sloppy in general.

The only place I differ is just in naming preference, in that I like `InflectionRuleChainCandidates` over `PossibleInflectionRuleChains` because it keeps the same suffix as the other types and "candidate" sounds more like a programming term (e.g. "release candidate").

---

The only other thing that comes to my mind with regard to the term "inflection" is "what about things that aren't strictly-speaking an inflection?" For example, text transformations can still occur in other ways, and I think I've probably mentioned that the extension 10ten does this for dates. I think I may have touched on this briefly in https://github.com/FooSoft/yomichan/issues/1410 a long time ago.

Overall, this last part of my comment goes beyond the scope of this change, but I wonder if perhaps eventually the idea of an "Inflection" should be further generalized to just a "Transformation" or something like that, and then potentially categorize it as "inflection" or "`<insert-something-else-here>`".

File: ext/data/templates/default-anki-field-templates.handlebars
Comment (StefanVukovic99) [reply ext/data/templates/default-anki-field-templates.handlebars:1]: If you mean those `anki-field-templates-upgrade-vXY.handlebars` files, I guess i need a new v24 file, with the merge conflict markers for replacing the conjugation template? Should I include something there for #434?

**Review by toasted-nutbread — APPROVED** [review]

File: types/ext/dictionary.d.ts (hunk 7 above)
Comment (djahandarie) [reply types/ext/dictionary.d.ts]: > I've kept PossibleInflectionRuleChain as a type for now because the source property is on it, and InflectionRuleChain as a string[]
@StefanVukovic99 Ok -- I think it's fine for now, but in the end the source should probably be on an Inflection Rule.

toasted-nutbread ideas all sound good so I think we can maybe do a further generalization in the future.

**Review by djahandarie — APPROVED** [review]
Thank you again, let's keep going :mechanical_arm:
